src/rulesets/rest/2022-05-25/property-rules.ts

```diff
@@ -185,6 +185,54 @@ const arrayWithItemsInResponse = new ResponseBodyRule({
   },
 });
 
+const requiredPropertiesDeclaredInRequestBody = new RequestRule({
+  name: "request schema properties",
+  rule: (requestAssertions) => {
+    requestAssertions.body.requirement(
+      "declare required properties in objects",
+      (body) => {
+        checkRequiredProperties((body.raw as any).schema);
+      },
+    );
+  },
+});
+
+const requiredPropertiesDeclaredInResponse = new ResponseBodyRule({
+  name: "response schema properties",
+  rule: (responseAssertions) => {
+    responseAssertions.body.requirement(
+      "declare required properties in objects",
+      (body) => {
+        checkRequiredProperties(body.raw.schema as OpenAPIV3.SchemaObject);
+      },
+    );
+  },
+});
+
+const checkRequiredProperties = (schema: OpenAPIV3.SchemaObject) => {
+  const schemas = [schema];
```

Comment:
Since we're only ever passing one schema, and you then operate on schemas independently in the while loop, why make this an array? Seems like it would be simpler to just operate on the passed in schema then wrap it like this.

I'm assuming there's a detail to this that I'm missing which requires an array?

Reply:
Conceptually this is doing a recursive check on all properties which are themselves objects, but an array is being used as a stack here to eliminate that recursion since TS isn't tail-call optimized.